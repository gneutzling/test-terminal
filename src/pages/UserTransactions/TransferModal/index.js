import React, { useState } from "react"
import { isNil } from "ramda"
import { ApolloConsumer } from "react-apollo"
import { makeStyles } from "@material-ui/core/styles"
import Button from "@material-ui/core/Button"
import TextField from "@material-ui/core/TextField"
import Typography from "@material-ui/core/Typography"
import Modal from "app/common/Modal"
import QUERY_TRANSACTIONS from "../query"

const useStyles = makeStyles(() => ({
  buttonsWrapper: {
    display: "flex",
    justifyContent: "flex-end",
    marginTop: 20,
  },
}))

const TransferModal = props => {
  const classes = useStyles()
  const [formState, setFormState] = useState({
    fromId: props.userId,
  })

  const handleChange = name => event => {
    setFormState({
      ...formState,
      [name]: event.target.value,
    })
  }

  return (
    <Modal renderOpenButton={<Button color="primary">Transfer ETH</Button>}>
      {({ closeModal }) => {
        return (
          <ApolloConsumer>
            {client => {
              const confirmTransaction = async () => {
                await client.query({
                  query: QUERY_TRANSACTIONS,
                  variables: {
                    user: formState.toId,
                  },
                })

                const sender = client.readQuery({
                  query: QUERY_TRANSACTIONS,
                  variables: { user: formState.fromId },
                })

                const requestedTransaction = sender.transactions.find(
                  ({ id }) => id === props.transaction.id
                )

                client.writeQuery({
                  query: QUERY_TRANSACTIONS,
                  variables: { user: formState.fromId },
                  data: {
                    transactions: sender.transactions.filter(
                      ({ id }) => id !== props.transaction.id
                    ),
                  },
                })

                const receiver = client.readQuery({
                  query: QUERY_TRANSACTIONS,
                  variables: { user: formState.toId },
                })

                client.writeQuery({
                  query: QUERY_TRANSACTIONS,
                  variables: { user: formState.toId },
                  data: {
                    transactions: receiver.transactions.concat([
                      requestedTransaction,
                    ]),
                  },
                })
              }

              return (
                <>
                  <Typography variant="h6">Transfer ETH</Typography>
                  <Typography variant="caption">
                    Transaction ID: {props.transaction.id}
                  </Typography>
                  <TextField
                    id="from-id"
                    label="From ID"
                    value={formState.fromId}
                    onChange={handleChange("fromId")}
                    margin="normal"
                    fullWidth
                    disabled
                  />
                  <TextField
                    id="to-id"
                    label="To ID"
                    onChange={handleChange("toId")}
                    margin="normal"
                    fullWidth
                  />

                  <div className={classes.buttonsWrapper}>
                    <Button color="secondary" onClick={closeModal}>
                      Cancel
                    </Button>
                    <Button
                      variant="contained"
                      color="primary"
                      onClick={confirmTransaction}
                      disabled={isNil(formState.toId)}
                    >
                      Confirm
                    </Button>
                  </div>
                </>
              )
            }}
          </ApolloConsumer>
        )
      }}
    </Modal>
  )
}

export default TransferModal
