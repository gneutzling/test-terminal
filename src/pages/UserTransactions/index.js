import React from "react"
import { pathOr } from "ramda"
import Query from "app/common/Query"
import QUERY_USER from "./query"

const UserTransactions = props => {
  return (
    <Query
      query={QUERY_USER}
      variables={{
        id: props.match.params.userId,
      }}
    >
      {({ data }) => {
        const exchangeBalances = pathOr([], ["user", "exchangeBalances"], data)

        return (
          <div>
            {exchangeBalances.map(balance => {
              return <div key={balance.id}>{balance.id}</div>
            })}
          </div>
        )
      }}
    </Query>
  )
}

export default UserTransactions
