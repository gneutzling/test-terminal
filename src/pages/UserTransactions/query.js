import gql from "graphql-tag"

export default gql`
  query User($id: ID!) {
    user(id: $id, first: 10) {
      id
      exchangeBalances {
        id
      }
      txs {
        id
      }
    }
  }
`
