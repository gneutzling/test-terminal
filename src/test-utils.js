import React from "react"
import { render } from "@testing-library/react"
import "jest-dom/extend-expect"

import RouterProvider from "app/lib/RouterProvider"

const AllTheProviders = ({ children }) => {
  return <RouterProvider>{children}</RouterProvider>
}

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options })

// re-export everything
export * from "@testing-library/react"

// override render method
export { customRender as render }
