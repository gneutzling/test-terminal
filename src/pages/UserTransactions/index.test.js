import React from "react"
import { MockedProvider } from "react-apollo/test-utils"
import wait from "waait"
import { render } from "app/test-utils"
import QUERY_TRANSACTIONS from "./query"
import UserTransactions from "./"

const build = ({ mocks, ...props }) =>
  render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <UserTransactions {...props} />
    </MockedProvider>
  )

const defaultMocks = [
  {
    request: {
      query: QUERY_TRANSACTIONS,
      variables: {
        user: "0x0000000000c90bc353314b6911180ed7e06019a9",
      },
    },
    result: {
      data: {
        transactions: [
          {
            ethAmount: "0.999799117276250112",
            fee: "0.003008422619687813",
            id: "207503",
            timestamp: 1561641980,
            tokenAmount: "1086.866722435225686668",
            tx:
              "0x12bd6f147df983c46a5c0a296f86efc42fc110223a6cdf494e9c36a5fd6dc138",
          },
        ],
      },
    },
  },
]

const defaultProps = {
  match: {
    params: {
      userId: "0x0000000000c90bc353314b6911180ed7e06019a9",
    },
  },
}

describe("UserTransactions", () => {
  it("matches snapshot", async () => {
    const { container } = build({ mocks: defaultMocks, ...defaultProps })
    await wait(0) // wait for response
    expect(container).toMatchSnapshot()
  })
})
