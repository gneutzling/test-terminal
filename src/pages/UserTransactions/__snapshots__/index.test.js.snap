// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserTransactions matches snapshot 1`] = `
<div>
  <div>
    <div>
      0x0cf0ee63788a0849fe5297f3407f701e122cc023-0x0000000000c90bc353314b6911180ed7e06019a9
    </div>
  </div>
</div>
`;
